Kubevirt: Enforce limits & requests by a configurable ratio

iholder101 opened the pull request:
This PR brings a mechanism to align Kubevirt `virt-launcher` Pods with ResourceQuotas that are applied to the namespace.

As an example, let's look at the following VMI:
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  labels:
    special: vmi-fedora
  name: vmi-fedora
spec:
  domain:
    devices:
      disks:
      - disk:
          bus: virtio
        name: containerdisk
      - disk:
          bus: virtio
        name: cloudinitdisk
      rng: {}
    resources:
      requests:
        memory: 1024M
  terminationGracePeriodSeconds: 0
  volumes:
  - containerDisk:
      image: quay.io/kubevirt/fedora-with-test-tooling-container-disk:20230115_2baf5303d 
    name: containerdisk
  - cloudInitNoCloud:
      userData: |-
        #cloud-config
        password: fedora
        chpasswd: { expire: False }
    name: cloudinitdisk
```

When this VMI is created, the following `virt-launcher` pod is created (some details are omitted for simplicity):
```yaml
kind: Pod
metadata:
  name: virt-launcher-vmi-fedora-lzzn6
  namespace: kubevirt-hyperconverged
spec:
  containers:
    name: compute
    resources:
      limits:
        devices.kubevirt.io/kvm: "1"
        devices.kubevirt.io/tun: "1"
        devices.kubevirt.io/vhost-net: "1"
      requests:
        cpu: 100m
        devices.kubevirt.io/kvm: "1"
        devices.kubevirt.io/tun: "1"
        devices.kubevirt.io/vhost-net: "1"
        ephemeral-storage: 50M
        memory: "1279755392"
```

As can be seen, this `virt-launcher` has only CPU and memory requests - but not limits. This means that if this VMI is being created in a namespace that has a ResourceQuota defined in it - the virt-launcher Pod won't be able to start. This now can be solved by the mechanism that's presented in this PR. 

To enable this mechanism, first a ratio between memory/CPU limits to request needs to be defined as an annotation in HCO object:
```yaml
kind: HyperConverged
metadata:
  annotations:
    kubevirt.io/cpu-limit-to-request-ratio: "2"
    kubevirt.io/memory-limit-to-request-ratio: "1.5"
```

In addition, a ResourceQuota needs to exist on the relevant namespace. As an example, it's possible to create the following object:
```yaml
apiVersion: v1
kind: ResourceQuota
metadata:
  name: test-rq 
spec:
  hard:
    limits.cpu: "200"
    limits.memory: "2000G"
```
Please take into account that if a ResourceQuota only sets a limit on `limits.cpu` or `limits.cpu`, then CPU/memory limits will be set accordingly. If multiple ResourceQuota exist within the relevant namespace, it takes only one of then to limit CPU/memory limits in order to enforce these limits.

When these annotations are enabled along with a ResourceQuota object, a mutating webhook that's targeted to virt-launcher pods will enforce limits on the pod. It would now look like the following:
```yaml
    resources:
      limits:
        cpu: 200m
        devices.kubevirt.io/kvm: "1"
        devices.kubevirt.io/tun: "1"
        devices.kubevirt.io/vhost-net: "1"
        memory: "1919633088"
      requests:
        cpu: 100m
        devices.kubevirt.io/kvm: "1"
        devices.kubevirt.io/tun: "1"
        devices.kubevirt.io/vhost-net: "1"
        ephemeral-storage: 50M
        memory: "1279755392"
```

Bear in mind that the new webhook would be active only if a request is set but a limit is not.

**Reviewer Checklist**
<!-- Check [Expectations from a PR](/CONTRIBUTING.md#expectations-from-a-pr) for the details -->

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [X] PR Message
- [X] Commit Messages
- [X] How to test
- [X] Unit Tests
- [ ] Functional Tests
- [X] User Documentation
- [X] Developer Documentation
- [X] Upgrade Scenario
- [X] Uninstallation Scenario
- [X] Backward Compatibility
- [X] Troubleshooting Friendly
  
**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Kubevirt: Enforce limits & requests by a configurable ratio
```

nunnatsa commented on pkg/webhooks/mutator/virt-launcher-mutator.go:
```diff
@@ -0,0 +1,117 @@
+	}
+}
+
+func (m *VirtLauncherMutator) Handle(_ context.Context, req admission.Request) admission.Response {
```
Please use a real parameter for the context. You'll need it below.

nunnatsa commented on pkg/webhooks/mutator/virt-launcher-mutator.go:
```diff
@@ -0,0 +1,117 @@
+	}
+	originalPod := launcherPod.DeepCopy()
+
+	hco, err := getHcoObject(m.cli, m.hcoNamespace)
```
please us the context here.

nunnatsa commented on pkg/webhooks/mutator/utils.go:
```diff
@@ -0,0 +1,35 @@
+	hcoutil "github.com/kubevirt/hyperconverged-cluster-operator/pkg/util"
+)
+
+func getHcoObject(cli client.Client, namespace string) (*v1beta1.HyperConverged, error) {
+	ctx := context.TODO()
```
Please define the context as the first function parameter and use it.

iholder101 replied:
Will do.
FYI, the other webhooks are implemented in the same way (context isn't being used).

iholder101 replied:
@nunnatsa The ns mutator also calls this function, but it dismisses the context parameter.
You can see [here](https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/pkg/webhooks/mutator/mutator.go#L97) how I'm simply using the same logic that already exists and following the current pattern.

Does this mean that the already code is wrong?
Should I also add the context for other webhooks as well?

nunnatsa replied:
If you don't mind to fix it, please do. You're refactoring this code anyway.

iholder101 replied:
:+1:

nunnatsa reviewed (commented):
Added three comments about context, but they are all the same one, actually.

iholder101 commented:
/cc @enp0s3 @fabiand @stu-gott @acardace

nunnatsa commented on pkg/webhooks/mutator/utils.go:
```diff
@@ -0,0 +1,34 @@
+	hcoutil "github.com/kubevirt/hyperconverged-cluster-operator/pkg/util"
+)
+
+func getHcoObject(cli client.Client, ctx context.Context, namespace string) (*v1beta1.HyperConverged, error) {
```
The golang convention is that the context is always the first parameter in the function. Could you please reorder?

nunnatsa commented on pkg/webhooks/mutator/utils.go:
```diff
@@ -0,0 +1,34 @@
+	}
+
+	err := cli.Get(ctx, client.ObjectKeyFromObject(hco), hco)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("HCO CR doesn't not exist, allow hcoNamespace deletion")
+			return nil, err
+		}
+
+		logger.Error(err, "failed getting HyperConverged CR")
+		return nil, err
+	}
```
Is that the right error handling and error message, now that it's a generic function?

nunnatsa commented on pkg/webhooks/mutator/virt-launcher-mutator.go:
```diff
@@ -0,0 +1,178 @@
+
+	hco, err := getHcoObject(m.cli, ctx, m.hcoNamespace)
+	if err != nil {
+		m.logErr(err, "cannot get hco object")
```
maybe 
```suggestion
		m.logErr(err, "cannot get the HyperConverged object")
```

nunnatsa commented on pkg/webhooks/mutator/virt-launcher-mutator_test.go:
```diff
@@ -0,0 +1,159 @@
+		Expect(err).ToNot(HaveOccurred())
+
+		resources := launcherPod.Spec.Containers[0].Resources
+		Expect(resources.Limits[v1.ResourceCPU].Equal(expectedResources.Limits[v1.ResourceCPU])).To(BeTrue())
+		Expect(resources.Requests[v1.ResourceCPU].Equal(expectedResources.Requests[v1.ResourceCPU])).To(BeTrue())
+		Expect(resources.Limits[v1.ResourceMemory].Equal(expectedResources.Limits[v1.ResourceMemory])).To(BeTrue())
+		Expect(resources.Requests[v1.ResourceMemory].Equal(expectedResources.Requests[v1.ResourceMemory])).To(BeTrue())
```
Isn't the gomega's `Equals` matcher work here?

iholder101 replied:
Unfortunately not :(
With Gomega's Equals I get errors that are similar to:
```
  [FAILED] Expected
      <resource.Quantity>: {
          i: {value: 150000000000, scale: -3},
          d: {Dec: nil},
          s: "150M",
          Format: "DecimalSI",
      }
  to equal
      <resource.Quantity>: {
          i: {value: 150, scale: 6},
          d: {Dec: nil},
          s: "150M",
          Format: "DecimalSI",
      }
```

Although the two quantities are effectively equal (to `150M`).

nunnatsa commented on pkg/webhooks/mutator/virt-launcher-mutator_test.go:
```diff
@@ -0,0 +1,159 @@
+	podNamespace = "fake-namespace"
+)
+
+var _ = Describe("virt-launcher webhook mutator", func() {
```
Please add error cases; e.g. no HyperConverged, negative ratio, zero ratio etc.

iholder101 replied:
Done!
Added tests for no HCO and invalid ratio. IMHO we don't need to test the library that parses the annotation as it's beyond this "unit's" scope.

nunnatsa reviewed (commented):
Added a few comments

nunnatsa commented on pkg/webhooks/mutator/virt-launcher-mutator.go:
```diff
@@ -0,0 +1,178 @@
+	"github.com/kubevirt/hyperconverged-cluster-operator/api/v1beta1"
+)
+
+// TODO: Is that needed?
```
This is a way to promise in compile time, that the `VirtLauncherMutator` type actually implements the `admission.Handler` interface.

iholder101 replied:
Makes sense! Thanks a lot!

iholder101 commented:
> @iholder101 Hey :) Thank you for the PR! A general question, I can't find the place where we filter the pods based on those that belong to the actual virt-launcher, can you please refer me to the place?

Gladly :)
[Here](https://github.com/kubevirt/hyperconverged-cluster-operator/pull/2206/files#diff-f1531df30500812ec42a33a52023c9ceca8a93d4972132fd241ad143b107c063R824) you can find the `WebhookDescription` object being defined. Please look at the following segment:
```go
ObjectSelector: &metav1.LabelSelector{
  MatchLabels: map[string]string{"kubevirt.io": "virt-launcher"},
},
```

This ensures that only virt-launcher pods are being handled.
This specific label is being added to every virt-launcher pod by Kubevirt, you can see it [here](https://github.com/kubevirt/kubevirt/blob/v0.59.0-alpha.2/pkg/virt-controller/services/template.go#L1360).

vladikr commented on pkg/webhooks/mutator/virt-launcher-mutator.go:
```diff
@@ -0,0 +1,177 @@
+func (m *VirtLauncherMutator) setResourceRatio(launcherPod *k8sv1.Pod, ratioStr string, resourceName k8sv1.ResourceName) error {
+	ratio, err := strconv.ParseFloat(ratioStr, 64)
+	if err != nil {
+		return fmt.Errorf("cann't parse %s to float: %v", ratioStr, err)
```
s/cann't/can't

vladikr commented on pkg/webhooks/mutator/virt-launcher-mutator.go:
```diff
@@ -0,0 +1,177 @@
+	}
+
+	if ratio <= 0 {
+		return fmt.Errorf("negative ratio (%v) is not supported", ratio)
```
I think these error messages should be more self-explanatory. The final error won't have any addition context.

vladikr commented on pkg/webhooks/mutator/virt-launcher-mutator.go:
```diff
@@ -0,0 +1,177 @@
+			launcherPod.Spec.Containers[i].Resources.Limits[resourceName] = m.multiplyResource(request, ratio)
+		}
+		if !requestExists && limitExists {
+			launcherPod.Spec.Containers[i].Resources.Requests[resourceName] = m.multiplyResource(limit, 1/ratio)
```
there is no need to set resource requests if only limits are provided. When users provide only the limits it means that requests will be equal to limits.

vladikr commented:
@iholder101 Didn't we intend to apply limits only to pods in namespaces that actually require this? (i.e. have resourceQuota)

iholder101 commented:
@vladikr I believe I've addressed all of your comments. Now the mechanism is enabled only if a ResourceQuota exists on the namespace that limits cpu/memory limits.

In addition, I've:
- Added documentation
- Added a unit test to ensure that the mechanism doesn't work if only limits are set
- Updated PR description

Can you PTAL?
FYI @nunnatsa

vladikr commented:
Look good to me. Thank @iholder101 
/lgtm

nunnatsa commented on docs/cluster-configuration.md:
````diff
@@ -1075,4 +1075,77 @@ The `tuningPolicy` feature can be enabled using the following patch:
+        memory: "1279755392"
+```
+
+Bear in mind that the new webhook would be active only if a request is set but a limit is not.
````
Not sure the user cares about the webhook. We can maybe write something like this:
"Bear in mind that if the limit is already set, HCO will not fix it according to the annotations." or something like this.

WDYT?

iholder101 replied:
:+1:

nunnatsa commented on pkg/components/components.go:
```diff
@@ -490,6 +490,11 @@ func GetClusterPermissions() []rbacv1.PolicyRule {
 			Resources: stringListToSlice("namespaces"),
 			Verbs:     stringListToSlice("get", "list", "watch", "patch", "update"),
 		},
+		{
+			APIGroups: emptyAPIGroup,
+			Resources: stringListToSlice("resourcequotas"),
+			Verbs:     stringListToSlice("get", "list", "watch"),
+		},
```
nit: The rback section for pods shares the same setting. it is possible to merge them, to get shorter CSV.

iholder101 replied:
:+1:

nunnatsa commented on pkg/webhooks/mutator/virt-launcher-mutator.go:
```diff
@@ -0,0 +1,234 @@
+}
+
+func (m *VirtLauncherMutator) setResourceRatio(launcherPod *k8sv1.Pod, ratioStr, annotationKey string, resourceName k8sv1.ResourceName) error {
+	ratio, err := strconv.ParseFloat(ratioStr, 64)
+	if err != nil {
+		return fmt.Errorf("%s can't parse ratio %s to float: %v. The ratio is the value of annotation key %s", launcherMutatorStr, ratioStr, err, annotationKey)
+	}
+
+	if ratio <= 0 {
+		return fmt.Errorf("%s doesn't support negative ratio: %v. The ratio is the value of annotation key %s", launcherMutatorStr, ratio, annotationKey)
+	}
+
```
What do you think about adding this validation to the validation webhook? Not sure who called first, but if the validation webhook is the first one, it's better to fail fast.

iholder101 replied:
Mutating webhooks are actually called first :) (and with a good reason - the validating webhooks get the final manifest version that would actually get into etcd). Therefore I think it's better to fail here.

nunnatsa commented on pkg/webhooks/mutator/virt-launcher-mutator.go:
```diff
@@ -0,0 +1,234 @@
+func (m *VirtLauncherMutator) setResourceRatio(launcherPod *k8sv1.Pod, ratioStr, annotationKey string, resourceName k8sv1.ResourceName) error {
+	ratio, err := strconv.ParseFloat(ratioStr, 64)
+	if err != nil {
+		return fmt.Errorf("%s can't parse ratio %s to float: %v. The ratio is the value of annotation key %s", launcherMutatorStr, ratioStr, err, annotationKey)
```
please use `%w` to embed the error in `fmt.Errorf`.

nunnatsa commented on pkg/webhooks/mutator/virt-launcher-mutator.go:
```diff
@@ -0,0 +1,234 @@
+	}
+
+	if ratio <= 0 {
+		return fmt.Errorf("%s doesn't support negative ratio: %v. The ratio is the value of annotation key %s", launcherMutatorStr, ratio, annotationKey)
```
same

nunnatsa commented on docs/cluster-configuration.md:
````diff
@@ -1075,4 +1075,77 @@ The `tuningPolicy` feature can be enabled using the following patch:
\ No newline at end of file
+```
+
+## Enforce CPU/memory limits on namespaces with ResourceQuotas
````
Great documentation. Very clear and helpful. :1st_place_medal:

nunnatsa commented:
/lgtm
/override coverage/coveralls

nunnatsa commented:
/test hco-e2e-image-index-sno-aws
/test hco-e2e-upgrade-index-aws
/test hco-e2e-upgrade-index-sno-aws
/test hco-e2e-upgrade-prev-index-aws
/test hco-e2e-upgrade-prev-index-sno-aws
/test okd-hco-e2e-image-index-aws
/test okd-hco-e2e-upgrade-index-aws
/test okd-hco-e2e-upgrade-index-gcp

nunnatsa commented:
/retest

iholder101 commented:
/retest

iholder101 commented:
/test hco-e2e-upgrade-prev-index-sno-azure
/test hco-e2e-upgrade-prev-index-sno-aws
/test hco-e2e-upgrade-index-sno-aws
/test hco-e2e-upgrade-index-aws
/test okd-hco-e2e-upgrade-index-aws
/test hco-e2e-upgrade-prev-index-aws
/test hco-e2e-upgrade-index-sno-azure
/test hco-e2e-upgrade-index-azure
/test hco-e2e-upgrade-prev-index-azure

iholder101 commented:
/test okd-hco-e2e-upgrade-index-gcp

orenc1 commented:
upgrade lanes failed due to an x509 certificate validation of the validating webhook upon HCO CR deletion:
`x509: certificate signed by unknown authority`
which resembles the bug we're opened against OLM a while ago:
https://bugzilla.redhat.com/show_bug.cgi?id=1907290

/retest

orenc1 commented on docs/cluster-configuration.md:
````diff
@@ -1075,4 +1075,77 @@ The `tuningPolicy` feature can be enabled using the following patch:
+    limits.cpu: "200"
+    limits.memory: "2000G"
+```
+Please take into account that if a ResourceQuota only sets a limit on `limits.cpu` or `limits.cpu`, then CPU/memory limits will be set accordingly. If multiple ResourceQuota exist within the relevant namespace, it takes only one of then to limit CPU/memory limits in order to enforce these limits.
````
did you mean this?
```suggestion
Please take into account that if a ResourceQuota only sets a limit on `limits.cpu` or `limits.memory`, then CPU/memory limits will be set accordingly. If multiple ResourceQuota exist within the relevant namespace, it takes only one of then to limit CPU/memory limits in order to enforce these limits.
```

orenc1 commented on pkg/webhooks/mutator/utils.go:
```diff
@@ -0,0 +1,34 @@
+	err := cli.Get(ctx, client.ObjectKeyFromObject(hco), hco)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("HCO CR doesn't not exist")
```
```suggestion
			logger.Info("HyperConverged CR does not exist")
```

orenc1 commented on pkg/webhooks/mutator/virt-launcher-mutator.go:
```diff
@@ -0,0 +1,234 @@
+	m.logInfo("Starting virt-launcher mutator handling")
+
+	if req.Operation != admissionv1.Create {
+		m.logInfo("not a pod creation - ignoring")
```
if we got here, i think it is an error in the webhook definition, since it is defined to forward the request to the webhook handler only on pods creation:
https://github.com/kubevirt/hyperconverged-cluster-operator/pull/2206/files#diff-a9655738a5e2e547cd22e5e3a3b56329022658b0ae27e94c49f54c18d12ce362R3853-R3861

iholder101 replied:
Yeah that's right. I just prefer to also have it here as a second line of defense. It also serves as "documentation", since the configuration is located in a different file that would be hard to find without context. In addition, the current namespace mutator does the same thing with deletions.

orenc1 commented on pkg/webhooks/mutator/virt-launcher-mutator.go:
```diff
@@ -0,0 +1,234 @@
+		memRatioStr = hco.Annotations[memoryLimitToRequestRatioAnnotation]
+	}
+
+	if enforceCpuLimits {
+		err := m.setResourceRatio(launcherPod, cpuRatioStr, cpuLimitToRequestRatioAnnotation, k8sv1.ResourceCPU)
+		if err != nil {
+			return err
+		}
+	}
```
why this block is not at the same block at lines 90-91 above?
same comment for the other block of:
`if enforceMemoryLimits {`

orenc1 commented on pkg/webhooks/mutator/virt-launcher-mutator.go:
```diff
@@ -0,0 +1,234 @@
+		return fmt.Errorf("%s can't parse ratio %s to float: %w. The ratio is the value of annotation key %s", launcherMutatorStr, ratioStr, err, annotationKey)
+	}
+
+	if ratio <= 0 {
```
i would even say that a ratio < 1 is not supported, because a limit that is smaller than the request makes no sense.

orenc1 commented on pkg/webhooks/mutator/virt-launcher-mutator.go:
```diff
@@ -0,0 +1,234 @@
+
+		if requestExists && !limitExists {
+			newQuantity := m.multiplyResource(request, ratio)
+			m.logInfo("Replacing old quantity (%s) with new quantity (%s) for pod %s/%s, UID: %s, accodring to ratio: %v",
```
i think it would be nice to specify in the log if the quantity that was replaced is memory or cpu.

orenc1 commented on pkg/webhooks/mutator/namespace_mutator.go:
```diff
@@ -7,23 +7,19 @@ import (
-	admittingDeletionMessage = "the namespace doesn't contain HyperConverged CR, admitting its deletion"
-	deniedDeletionMessage    = "HyperConverged CR is still present, please remove it before deleting the containing namespace"
+	admittingDeletionMessage = "the hcoNamespace doesn't contain HyperConverged CR, admitting its deletion"
+	deniedDeletionMessage    = "HyperConverged CR is still present, please remove it before deleting the containing hcoNamespace"
```
**IMPORTANT**
you changed here the error message, replacing `namespace` to `hcoNamespace`.
the delete namespace test checks the exact error message, thus you need to update it also there:
https://github.com/kubevirt/hyperconverged-cluster-operator/blob/43945370a6322b92404a817d6cc8e99292552ab7/hack/test_delete_ns.sh#L41

iholder101 replied:
Thanks a lot!!!
For the record, I think it's a bad idea to expect to the full error message. As we saw, it leads to surprising errors and enforces the developer to duplicate the error code in two places. I think it should be fixed in future PRs :+1:

orenc1 replied:
it's just because you've decided to change `namespace` to `hcoNamespace` :)
this error message was the same for years

iholder101 commented:
Thanks a lot @orenc1!
I've addressed your comments above

orenc1 commented on pkg/webhooks/mutator/virt-launcher-mutator.go:
```diff
@@ -0,0 +1,230 @@
+		return fmt.Errorf("%s can't parse ratio %s to float: %w. The ratio is the value of annotation key %s", launcherMutatorStr, ratioStr, err, annotationKey)
+	}
+
+	if ratio <= 1 {
```
i think `ratio == 1` should be allowed, although it is not so common to have the requests the same as limits, but it is valid configuration.

iholder101 replied:
:+1:

iholder101 commented:
/retest

orenc1 commented:
closing and reopening to release the stuck lanes
/close

orenc1 commented:
/reopen

orenc1 commented:
/retest

orenc1 commented:
since it is a temporary solution for the resource limits problem in VirtualMachines, and it is going to get a permanent fix in kubevirt/kubevirt side, i'm ok with disregarding the coverage decrease.
/override coverage/coveralls
/lgtm
/approve

iholder101 commented:
/cherrypick release-1.6

iholder101 commented:
/cherrypick release-1.8
